l1/l1.go

```diff
@@ -0,0 +1,134 @@
+package l1
+
+import (
+	"container/list"
+	"context"
+	"fmt"
+
+	"github.com/NethermindEth/juno/blockchain"
+	"github.com/NethermindEth/juno/core"
+	"github.com/NethermindEth/juno/core/felt"
+	"github.com/NethermindEth/juno/l1/contract"
+	"github.com/NethermindEth/juno/service"
+	"github.com/NethermindEth/juno/utils"
+	"github.com/ethereum/go-ethereum/core/types"
+	"github.com/ethereum/go-ethereum/event"
+)
+
+var (
+	ConfirmationPeriod uint64 = 1
+
+	base16 = 16
+)
+
+//go:generate mockgen -destination=../mocks/mock_subscriber.go -package=mocks github.com/NethermindEth/juno/l1 Subscriber
+type Subscriber interface {
+	WatchLogStateUpdate(ctx context.Context, sink chan<- *contract.StarknetLogStateUpdate) (event.Subscription, error)
+	WatchHeader(ctx context.Context, ch chan<- *types.Header) (event.Subscription, error)
+}
+
+type Client struct {
+	l1      Subscriber
+	l2Chain *blockchain.Blockchain
+	log     utils.SimpleLogger
+	q       *list.List
+}
+
+var _ service.Service = (*Client)(nil)
+
+func NewClient(l1 Subscriber, chain *blockchain.Blockchain, log utils.SimpleLogger) *Client {
+	return &Client{
+		l1:      l1,
+		l2Chain: chain,
+		log:     log,
+		q:       list.New(),
+	}
+}
+
+func (c *Client) Run(ctx context.Context) error {
+	subscribeCtx, subscribeCtxCancel := context.WithCancel(ctx)
+	defer subscribeCtxCancel()
+	buffer := 128
+
+	headerChan := make(chan *types.Header, buffer)
+	subHeader, err := c.l1.WatchHeader(subscribeCtx, headerChan)
+	if err != nil {
+		return fmt.Errorf("subscribe to L1 headers: %w", err)
+	}
+	defer subHeader.Unsubscribe()
+
+	logStateUpdateChan := make(chan *contract.StarknetLogStateUpdate, buffer)
+	subLogStateUpdate, err := c.l1.WatchLogStateUpdate(subscribeCtx, logStateUpdateChan)
+	if err != nil {
+		return fmt.Errorf("subscribe to L1 state updates: %w", err)
+	}
+	defer subLogStateUpdate.Unsubscribe()
+
+	for {
+		select {
+		case err := <-subLogStateUpdate.Err():
+			return err
+		case err := <-subHeader.Err():
+			return err
+		case <-ctx.Done():
+			return nil
+		case header := <-headerChan:
+		Outer:
+			for {
```

Comment:
I think the loop should terminate (see `break Outer` in the `default` case). I added a test and it passes, but maybe I'm missing something?

I agree the order of messages doesn't matter in practice. However, it does allow us to make the tests deterministic. That's why I ended up nesting the `select`s. 

I think waiting 2 epochs for L1 is maybe a bit much? The overwhelming majority of [past reorgs](https://etherscan.io/blocks_forked?p=1) appear to be single-block reorgs. I couldn't find a non-single block reorg in the last 54 days. What do you think?

Reply:
> I agree the order of messages doesn't matter in practice. However, it does allow us to make the tests deterministic. That's why I ended up nesting the selects.

I would avoid doing that just for the sake of easier testing. Could we have updated the test instead?

Reply:
> doesnt that just jump to start of the loop again?

It should break out of the `Outer` loop. [This test](https://github.com/NethermindEth/juno/blob/37a5d711173450f0cc3bb1f2e7f348d1217a548a/l1/l1_test.go#L122) wouldn't succeed if it didn't behave as expected.

> Could we have updated the test instead?

I don't think so. I feel the tests should be agnostic to the order the messages are delivered. With the nested selects, we only rely on the assumption: "if we receive logs, then we also receive a header around the same time."

Reply:
> we only rely on the assumption: "if we receive logs, then we also receive a header around the same time."

Also we rely on events coming before the new header, dont we? Nested select is a bit racey, although since headers are pretty frequent it might not make a difference in the end.

Assume you get a header for height `N`, but events for that height `N` arrive after some time. So nested select will exit thru the default case since the events are not in the channels yet and we will only be processing the events after the next header for `N+1`.

Reply:
Not really. Other than making the flat-select work

Reply:
The flat-select has the same issue, right?

Reply:
Maybe, I might have not really thought this thru. We dont really need to fix this, it is not an issue at all. It is just an implicit `+1` to the confirmation period.